src/rust/src/lib.rs

```diff
@@ -667,118 +601,118 @@ fn get_rustc_version() -> String {
+        }
+    };
 }
 
-fn get_soroban_xdr_bindings_pkg_versions() -> VersionStringPair {
-    VersionStringPair {
-        curr: soroban_curr::soroban_env_host::VERSION.xdr.pkg.to_string(),
-        #[cfg(feature = "soroban-env-host-prev")]
-        prev: soroban_prev::soroban_env_host::VERSION.xdr.pkg.to_string(),
-        #[cfg(not(feature = "soroban-env-host-prev"))]
-        prev: "".to_string(),
-    }
-}
+// NB: this list should be in ascending order. Out of order will cause
+// an assert to fail in the by-protocol-number lookup function below.
+const HOST_MODULES: &'static [HostModule] = &[
+    proto_versioned_functions_for_module!(soroban_p21),
+    proto_versioned_functions_for_module!(soroban_p22),
+];
 
-fn get_soroban_xdr_bindings_git_versions() -> VersionStringPair {
-    VersionStringPair {
-        curr: soroban_curr::soroban_env_host::VERSION.xdr.rev.to_string(),
-        #[cfg(feature = "soroban-env-host-prev")]
-        prev: soroban_prev::soroban_env_host::VERSION.xdr.rev.to_string(),
-        #[cfg(not(feature = "soroban-env-host-prev"))]
-        prev: "".to_string(),
+fn get_host_module_for_protocol(
+    config_max_protocol: u32,
+    ledger_protocol_version: u32,
+) -> Result<&'static HostModule, Box<dyn std::error::Error>> {
+    if ledger_protocol_version > config_max_protocol {
+        return Err(Box::new(soroban_curr::contract::CoreHostError::General(
+            "protocol exceeds configured max",
+        )));
     }
-}
-
-fn get_soroban_xdr_bindings_base_xdr_git_versions() -> VersionStringPair {
-    let curr = match soroban_curr::soroban_env_host::VERSION.xdr.xdr {
-        "next" => soroban_curr::soroban_env_host::VERSION
-            .xdr
-            .xdr_next
-            .to_string(),
-        "curr" => soroban_curr::soroban_env_host::VERSION
-            .xdr
-            .xdr_curr
-            .to_string(),
-        _ => "unknown configuration".to_string(),
-    };
-    #[cfg(feature = "soroban-env-host-prev")]
-    let prev = match soroban_prev::soroban_env_host::VERSION.xdr.xdr {
-        "next" => soroban_prev::soroban_env_host::VERSION
-            .xdr
-            .xdr_next
-            .to_string(),
-        "curr" => soroban_prev::soroban_env_host::VERSION
-            .xdr
-            .xdr_curr
-            .to_string(),
-        _ => "unknown configuration".to_string(),
-    };
-    #[cfg(not(feature = "soroban-env-host-prev"))]
-    let prev = "".to_string();
-    VersionStringPair { curr, prev }
-}
-
-impl std::fmt::Display for rust_bridge::BridgeError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{:?}", self)
+    // Each host's max protocol implies a min protocol for the _next_
+    // host in the list. The first entry's min protocol is 0.
+    let mut curr_min_proto = 0;
+    for curr in HOST_MODULES.iter() {
+        assert!(curr_min_proto <= curr.max_proto);
```

Comment:
This method is called in the apply path, in `compute_transaction_resource_fee`, and in `compute_rent_fee`, so this assert makes me nervous. 

1. Do we validate `HOST_MODULES` isn't misconfigured anywhere on startup? It doesn't look like it.
2. Should we return an error here just to be safe?

Reply:
I will add a check on startup.